R/utils-latency.R

```diff
@@ -0,0 +1,191 @@
+    to_keep <- to_keep | grepl(col_name, relevant_columns)
+  }
+  relevant_columns <- relevant_columns[to_keep]
+  if (length(relevant_columns) == 0) {
+    cli::cli_abort("There is no column(s) {terms_used}.",
+      current_column_names = names(new_data),
+      class = "epipredict_adjust_latency_nonexistent_column_used",
+      call = call
+    )
+  }
+  # this pulls text that is any number of digits between two _, e.g. _3557_, and
+  # converts them to an integer
+  shift_amounts <- stringr::str_match(relevant_columns, "_(\\d+)_") %>%
+    `[`(, 2) %>%
+    as.integer()
+
+  shift_cols <- dplyr::tibble(
+    original_name = relevant_columns,
+    shifts = shift_amounts
+  )
+  if (is.null(latency)) {
+    shift_cols %<>%
+      rowwise() %>%
+      # add the latencies to shift_cols
+      mutate(latency = get_latency(
+        new_data, as_of, original_name, shifts, sign_shift
+      )) %>%
+      ungroup()
+  } else if (length(latency) > 1) {
+    shift_cols %<>% rowwise() %>%
+      mutate(latency = unname(latency[purrr::map_lgl(
+        names(latency),
+        \(x) grepl(x, original_name)
+      )])) %>%
+      ungroup()
+  } else {
+    shift_cols %<>% mutate(latency = latency)
+  }
+
+  # add the updated names to shift_cols
+  shift_cols %<>%
+    mutate(
+      effective_shift = shifts + abs(latency)
+    ) %>%
+    mutate(
+      new_name = adjust_name(prefix, original_name, effective_shift)
+    )
+  info %<>% select(variable, type, role)
+  shift_cols <- left_join(shift_cols, info, by = join_by(original_name == variable))
+  if (length(unique(shift_cols$role)) != 1) {
+    cli::cli_error("not all roles are the same!",
+      shift_cols = shift_cols
+    )
+  }
+  return(shift_cols)
+}
+
+
+#' extract the as_of, and make sure there's nothing very off about it
+#' @keywords internal
+set_asof <- function(new_data, info) {
+  original_columns <- info %>%
+    filter(source == "original") %>%
+    pull(variable)
+  # make sure that there's enough column names
+  if (length(original_columns) < 3) {
+    cli::cli_abort(glue::glue(
+      "The original columns of `time_value`, ",
+      "`geo_value` and at least one signal. The current colums are \n",
+      paste(capture.output(object$info), collapse = "\n\n")
+    ))
+  }
+  # the source data determines the actual time_values
+  # these are the non-na time_values;
+  time_values <- new_data %>%
+    select(all_of(original_columns)) %>%
+    drop_na() %>%
+    pull(time_value)
+  if (length(time_values) <= 0) {
+    rlang::abort("the `time_value` column of `new_data` is empty")
+  }
+  as_of <- attributes(new_data)$metadata$as_of
+  max_time <- max(time_values)
+  # make sure the as_of is sane
+  if (!inherits(as_of, class(time_values))) {
+    rlang::abort(glue::glue(
+      "the data matrix `as_of` value is {as_of}, ",
+      "and not a valid `time_type` with type ",
+      "matching `time_value`'s type of ",
+      "{typeof(new_data$time_value)}."
+    ))
```

Comment:
Is there a reason for rlang versions here and cli elsewhere?

Reply:
nope, not sure I put much thought into it. The package is also in a split state; is it alright if I just find-replace all instances of `rlang::abort` with `cli::cli_abort`, or is there some reason there are both generally?

Reply:
At the beginning, we used `{rlang}`, but I've been trying to have all new code use `{cli}` (with a plan to find and replace all the old stuff eventually). I thought there was an issue, but I can't find it.

Reply:
You can't generally just find-replace.  See [this migration info](https://rlang.r-lib.org/reference/topic-condition-formatting.html#transitioning-from-abort-to-cli-abort-).

Reply:
ah, they have a guide! Thanks Logan. I'll go over the changed ones and make sure they're not broken. Thankfully none of them broke the tests, but I'm guessing we don't have super strict `expect_error` tests for every single one